design.md

```diff
@@ -0,0 +1,372 @@
+        "taco": 200,
+        "chips": 300
+}|;
+from raw | records:=collect(this),seqno:=max(seqno),done:=or(value.done) by txn:=value.txn
+| done==true
+| cut this:={
+    customerID:customerIDs[records[0].value.row.customer],
+    menuID:menuIDs[records[0].value.row.product],
+    qty: records[0].value.row.qty,
+    total: records[0].value.row.qty * records[1].value.row.price
+  }
+```
+You can see the processed record by running this:
+```
+zapi query -I demo/etl.zed
+```
+and we get...
+```
+{customerID:1,menuID:200,qty:2,total:3.98}
+```
+Zinger sync would then sync this results --- which here is a single record
+but would more generally be multiple records --- by wrapping it in Kafka meta
+info and updating the cursor, e.g., as follows:
+```
+{
+  seqno:3,
+  kafka:{topic:"order",offset:1},
+  row:{customerID:1,menuID:200,qty:2,total:3.98}
+}
+{seqno:2}(=cursor)
+```
+The cursor is now at 2 since there still is pending data with
+a `seqno` of 2.
+
+
+> NOTE this encapsulation will automatically run be `zinger sync` once
+> we add this capability.
+
+To illustrate the steps here, you can manually load this data into
+"staging" using:
+```
+zapi load -use staging@main demo/staging-1.zson
+```
+
+Next suppose new data arrives that complete the pending one.  Load it with:
+```
+zapi load -use raw@main demo/consume-2.zson
+```
+
+Now, suppose zinger restarts in this sitaution.  Here are the steps needed to
+merge the new transaction into "staging".
+
+First, we need to find the current cursor stored in "staging":
+```
+zapi query "from staging | is(type(cursor)) | max(seqno)"
+{max:2}
+```
+Ok, it's '2'.  Now we get all the transaction data from "raw" greater than 2
+and we get:
+```
+{seqno:2,kafka:{topic:"order",offset:2},value:{txn:10001,done:false,row:{customer:"bob",product:"burrito",qty:1}}}
+{seqno:3,kafka:{topic:"menu",offset:1},value:{txn:10000,done:true,row:{product:"taco",price:1.99}}}
+{seqno:4,kafka:{topic:"menu",offset:2},value:{txn:10001,done:true,row:{product:"burrito",price:5.99}}}
+```
+Note that seqno 3 has already been processed.  We can get a list of sequence numbers
+already processed from "staging", e.g.,
+```
+zapi query "from staging | not is(type(cursor)) | seqno >= 2 | cut seqno"
+```
+and we can do an anti-join with the "raw" transactions to get just the records
+that we want to process.  We'll put this in `demo/update.zed`:
+```
+from (
+  raw => seqno >= 2 | sort seqno;
+  staging => not is(type(cursor)) | seqno >= 2 | cut seqno | sort seqno;
+)
+| anti join on seqno=seqno
+```
+Run this with
+```
+zapi query -I demo/update.zed
+```
+And we get exactly the records for the pending transaction:
+```
+{seqno:2,kafka:{topic:"order",offset:2},value:{txn:10001,done:false,row:{customer:"bob",product:"burrito",qty:1}}}
+{seqno:4,kafka:{topic:"menu",offset:2},value:{txn:10001,done:true,row:{product:"burrito",price:5.99}}}
+```
+Now we can put together the anti-join step with the ETL step in `demo/update-etl.zed`:
```

Comment:
```suggestion
Now we can put together the anti join step with the ETL step in `demo/update-etl.zed`:
```

Reply:
But that's a noun modifier and should have a dash!